File: DesignPatterns/01_GOF23/DP_examples/Go/01_Creational/Factory/Product.go
package Factory

// IShape 产品接口
type IShape interface {
	Draw()
	Fill()
}

// ShapeKind 产品类型枚举
type ShapeKind int

const (
	UNKNOWN_SHAPE ShapeKind = iota
	CIRCLE
	SQUARE
	RECTANGLE
	BIG_CIRCLE
)

// Circle 圆形产品
type Circle struct{}

func (c *Circle) Draw() {
	println("Draw a Circle")
}

func (c *Circle) Fill() {
	println("Fill red")
}

// Square 正方形产品
type Square struct{}

func (s *Square) Draw() {
	println("Draw a Square")
}

func (s *Square) Fill() {
	println("Fill Green")
}

// Rectangle 矩形产品
type Rectangle struct{}

func (r *Rectangle) Draw() {
	println("Draw a Rectangle")
}

func (r *Rectangle) Fill() {
	println("Fill Blue")
}

// UnknownShape 未知形状产品
type UnknownShape struct{}

func (u *UnknownShape) Draw() {}

func (u *UnknownShape) Fill() {}

File: DesignPatterns/01_GOF23/DP_examples/Python/01_Creational/Factory/Product.py
from abc import ABC, abstractmethod

# 产品接口
class IShape(ABC):
    @abstractmethod
    def draw(self):
        pass
    
    @abstractmethod
    def fill(self):
        pass

# 产品类型枚举
class ShapeKind:
    NoneType = 0
    Circle = 1
    Square = 2
    Rectangle = 3
    BigCircle = 4

# 圆形产品
class Circle(IShape):
    def draw(self):
        print("Draw a Circle")
    
    def fill(self):
        print("Fill red")

# 正方形产品
class Square(IShape):
    def draw(self):
        print("Draw a Square")
    
    def fill(self):
        print("Fill Green")

# 矩形产品
class Rectangle(IShape):
    def draw(self):
        print("Draw a Rectangle")
    
    def fill(self):
        print("Fill Blue")

# 未知形状产品
class UnknownShape(IShape):
    def draw(self):
        pass
    
    def fill(self):
        pass
